### ISSUE TITLE
[Accordions] Accordions content display on initial load

### ISSUE THREAD

**Issue by juliaelman** [uswds/uswds#1465]
The accordion is set to display by default and causes a "flickering" of the content on initial load. This occurs because the "show/hide" functionality is set by `aria-hidden` attributes being set via JavaScript. If JavaScript is enabled, the content should be hidden **by default**. If JavaScript is disabled, the content should not be hidden to make it accessible.

Refs https://github.com/18F/web-design-standards/pull/1418#issuecomment-245741080

**Comment by shawnbot** [comment]
Okay, so there appear to be a couple of distinct "phases" to account for once an accordion is in the DOM:

Phase  | CSS | JS | Screen<br>readers | Sighted<br>users
:--- | :--- | :--- | :--- | :---
1. Loading | ❌  | ❌  | 🗣   | 🙈 
2. Loading | ✔️   | ❌  | 🗣   | 🙈  
3. Loaded! | ✔️   | ✔️   | 🗣   | 👀  

1. CSS and JavaScript haven't loaded yet. In this case, we want the content visible to screen readers _only_.
1. CSS has loaded, but JavaScript hasn't. Same as above, so our CSS shouldn't touch it yet.
1. JavaScript has loaded, so we can _explicitly_ hide the content from sighted users unless `aria-expanded="true"` on a given tab.

Here's the catch, though: **screen readers _should_ skip over closed tabs if JavaScript is loaded**. So let's redo that table for a given tab and include a column for whether it's open (`aria-expanded="true"`) or not:

Phase  | CSS | JS | Open? | Screen<br>readers | Sighted<br>users
:--- | :--- | :--- | :--- | :--- | :---
1a. Loading | ❌  | ❌  | ➕  | 🗣   | 👀 
2a. Loading | ✔️   | ❌  | ➕   | 🗣   | 👀 
3a. Loaded! | ✔️   | ✔️   | ➕   | 🗣   | 👀  
1b. Loading | ❌  | ❌  | ➖  | 🗣   | 🙈 
2b. Loading | ✔️   | ❌  | ➖   | 🗣   | 🙈  
3b. Loaded! | ✔️   | ✔️   | ➖   | 🙊   | 🙈 

In other words:
* Open tabs should _always_ be visible _and_ readable.
* Closed tabs should _always_ be invisible, and readable _only_ until (if) JavaScript loads.

Paging Dr. @nickbristow: does this sound right?

**Comment by nickbristow** [comment]
Ideally everything on the page is correct as it loads, something is hidden has the proper aria tags. If javascript changes the visible state of something, the aria is updated then. 

Tabs loaded open, should say they are expanded and vice versa.

**Comment by shawnbot** [comment]
Okay, thanks @nickbristow. Here's what I'm thinking:

Phase  | CSS | JS | Open? | Screen<br>readers | Sighted<br>users | Solution
:--- | :--- | :--- | :--- | :--- | :--- | :---
1a. Loading | ❌  | ❌  | ➕  | 🗣   | 👀 | visible by default
2a. Loading | ✔️   | ❌  | ➕   | 🗣   | 👀 | visible by default
3a. Loaded! | ✔️   | ✔️   | ➕   | 🗣   | 👀 | visible by default
1b. Loading | ❌  | ❌  | ➖  | 🗣   | 🙈 | [hide with CSS]
2b. Loading | ✔️   | ❌  | ➖   | 🗣   | 🙈 | [hide with CSS]
3b. Loaded! | ✔️   | ✔️   | ➖   | 🙊   | 🙈 | set HTML `hidden`

There are a bunch of different states to account for here, but in plain English:

1. Open panels:
  1. should _always_ be screen-reader-accessible and visible to sighted users.
1. Closed panels:
  1. should be screen reader-accessible _until JS loads_; and
  1. should _always_ be [invisible][hide with CSS] to sighted users.

The trickiest scenario is the closed panel before CSS or JS has loaded. Aside from inline styles and the `hidden` attribute (which would render them inaccessible if JS doesn't load, or until it does), there's no way to hide these elements visually until our CSS has loaded. I think we should ignore that case, since our CSS is blocking rendering of the page anyway.

To achieve the effect of visually hidden but "inert" (pre-JS) accordions being invisible but screen-reader-accessible, we could [hide all panels with CSS][hide with CSS] that _aren't_ immediately preceded by an expanded button:

```css
[aria-controls]:not([aria-expanded=true]) + .usa-accordion-content {
  /* position offscreen, etc. */
}
```

Or, because [`:not()` isn't supported by IE8](http://caniuse.com/#feat=css-sel3), hide all `.usa-accordion-content` elements by default, and specifically show the ones that _are_ immediately preceded by an expanded button:

```css
.usa-accordion .usa-accordion-content {
  /* position offscreen, etc. */
}

.usa-accordion [aria-expanded=true] + .usa-accordion-content {
  /* undo absolute positioning, display visually */
}
```

In either case, the JS can just manage the `aria-expanded` state of the buttons and `hidden` on the corresponding panels. So when it loads, it looks for `aria-expanded` on the buttons and updates the panels accordingly (doing nothing if `aria-expanded="true"`, setting `aria-expanded="false"` on the button and `hidden` on the panel otherwise), and any time a button is clicked it flips the value of `aria-expanded` and applies the same logic.

I _think_ that covers all the cases, but if there are holes in my plan I'd love to hear them!

[hide with CSS]: https://snook.ca/archives/html_and_css/hiding-content-for-accessibility

**Comment by maya** [comment]
Closing, flickering fixed in #1512.
